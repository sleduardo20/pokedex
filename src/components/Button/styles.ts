import { motion, MotionProps } from 'framer-motion';
import styled, { css, DefaultTheme } from 'styled-components';

type ContainerProps = {
  bg?: 'primary' | 'secondary' | 'darkBlue';
} & MotionProps;

const modifiers = {
  primary: (theme: DefaultTheme) => css`
    background: ${theme.colors.primary};
    color: ${theme.colors.white};
  `,
  secondary: (theme: DefaultTheme) => css`
    background: ${theme.colors.secondary};
    color: ${theme.colors.darkBlue};
  `,
  darkBlue: (theme: DefaultTheme) => css`
    background: ${theme.colors.darkBlue};
    color: ${theme.colors.white};
  `,
};

export const Container = styled(motion.button)<ContainerProps>`
  ${({ theme, bg }) => css`
    width: 24.3rem;
    height: 3.8rem;
    padding: ${theme.spacings.small};
    display: flex;
    justify-content: center;
    align-items: center;
    border-radius: ${theme.spacings.small};
    outline: none;
    border: none;
    font-size: ${theme.font.size.large};
    font-weight: ${theme.font.medium};
    text-transform: uppercase;

    cursor: pointer;
    transition: filter 0.2s ease-in-out;

    &:hover {
      filter: brightness(105%);
    }

    ${!!bg && modifiers[bg](theme)};
  `}
`;
