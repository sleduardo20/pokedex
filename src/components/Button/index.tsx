import { ButtonHTMLAttributes } from 'react';

import * as S from './styles';

export interface ButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  bg?: 'primary' | 'secondary' | 'darkBlue';
}

const Button: React.FC<ButtonProps> = ({
  bg = 'darkBlue',
  children,
  ...rest
}) => {
  return (
    <S.Container bg={bg} {...rest}>
      {children}
    </S.Container>
  );
};

export default Button;
