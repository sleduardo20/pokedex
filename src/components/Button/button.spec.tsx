import { screen } from '@testing-library/react';
import { renderWithTheme } from 'util/renderWithTheme';

import Button from '.';

describe('<Button/>', () => {
  it('should be able render Button correctly', () => {
    const { container } = renderWithTheme(
      <Button>
        <span>Buscar</span>
      </Button>,
    );

    expect(screen.getByRole('button')).toBeInTheDocument();
    expect(screen.getByText('Buscar')).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should be able render background color primary', () => {
    renderWithTheme(
      <Button bg="primary">
        <span>Entrar</span>
      </Button>,
    );

    expect(screen.getByRole('button')).toHaveStyle({ background: '#CC0000' });
  });

  it('should be able render background color secondary', () => {
    renderWithTheme(
      <Button bg="secondary">
        <span>Entrar</span>
      </Button>,
    );

    expect(screen.getByRole('button')).toHaveStyle({ background: '#FFCB05' });
  });

  it('should be able render background color darkBlue', () => {
    renderWithTheme(
      <Button bg="darkBlue">
        <span>Entrar</span>
      </Button>,
    );

    expect(screen.getByRole('button')).toHaveStyle({ background: '#003A70' });
  });
});
