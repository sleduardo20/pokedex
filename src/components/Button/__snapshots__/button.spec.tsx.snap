// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button/> should be able render Button correctly 1`] = `
.c0 {
  width: 24.3rem;
  height: 4.8rem;
  padding: 0.4rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 0.4rem;
  outline: none;
  border: none;
  font-size: 1.6rem;
  font-weight: 300;
  text-transform: uppercase;
  color: #FFFFFF;
  background: #003A70;
}

<button
  class="c0"
>
  <span>
    Buscar
  </span>
</button>
`;
